### PR TITLE
iscsi machine networks

### PR DESCRIPTION
**Pull request by adriengentil**
- **MGMT-18513: machine networks must not be part of iSCSI subnet**
- **MGMT-18513: Add machine networks validation when iSCSI boot disk is in use**

### CODE REVIEW

File: internal/host/validator.go [hunk 1]
```diff
@@ -1829,3 +1831,62 @@ func (v *validator) inventoryHasIP(inventory *models.Inventory, ipAddress string
...
+			machineNetworkPrefix, err = netip.ParsePrefix(string(machineNetwork.Cidr))
+			if err != nil {
+				return true
+			}
+			return machineNetworkPrefix.Contains(nicIP.Addr())
+		})
+		return found
+	})
+
+	if err != nil {
```
Comment (rccrdpccl) [inline internal/host/validator.go:1882]: where is this error produced? seems to be from `network.FindInterfaceByIPString` which it's already handled

File: internal/host/validations_test.go [hunk 2]
```diff
@@ -2339,4 +2339,255 @@ var _ = Describe("Validations test", func() {
...
+					{
+						Cidr:      "192.168.2.1/24",
+						ClusterID: clusterID,
+					},
+					{
+						Cidr:      "2001:0db8:85a3::/48",
+						ClusterID: clusterID,
+					},
+				},
+				[]*models.Interface{
```
Comment (rccrdpccl) [inline internal/host/validations_test.go:2581]: Is it worth adding a test case with an unparseable IP prefix? For both NIC and machine network?

Comment (rccrdpccl) [reply internal/host/validations_test.go:2581]: Actually if test separately and additionally only the function it will be way easier, and we could add all these edge cases

File: internal/host/validator.go [hunk 3]
```diff
@@ -1829,3 +1831,62 @@ func (v *validator) inventoryHasIP(inventory *models.Inventory, ipAddress string
...
+
+	nic, err := network.FindInterfaceByIPString(installationDisk.Iscsi.HostIPAddress, c.inventory.Interfaces)
+	if err != nil {
+		v.log.WithError(err).Warn("Cannot find network interface associated to iSCSI host IP address")
+		return ValidationError, "Cannot find network interface associated to iSCSI host IP address"
+	}
+
+	// Look if one of the IP set on the interface connected to the iSCSI disk belong to machine network CIDRs
+	ips := append(nic.IPV4Addresses, nic.IPV6Addresses...)
+	_, found := lo.Find(ips, func(ip string) bool {
```
Comment (rccrdpccl) [inline internal/host/validator.go:1864]: Maybe we could extract this function to make it both clearer to the reader and more testeable (as commented in the test we could test edge cases in the function test, and more generic tests in the validation test)

File: swagger.yaml [hunk 4]
```diff
@@ -6503,6 +6503,7 @@ definitions:
       - 'no-skip-installation-disk'
       - 'no-skip-missing-disk'
       - 'no-ip-collisions-in-network'
+      - 'iscsi-host-network-interface-does-not-belong-to-machine-cidr'
```
Comment (eifrach) [inline swagger.yaml:6506]: let make it a bit shorter -- It still OK leaving it as is 
also better you have a validation be a `True` if it's good and `False` when it fails - just a bit easier to understand when seen/debugging 

@AlonaKaplan what to you think ?
```suggestion
      - 'iscsi-net-belong-to-machine-cidr'
```

File: internal/host/validator.go [hunk 5]
```diff
@@ -1829,3 +1831,62 @@ func (v *validator) inventoryHasIP(inventory *models.Inventory, ipAddress string
...
+		_, found := lo.Find(c.cluster.MachineNetworks, func(machineNetwork *models.MachineNetwork) bool {
+			var machineNetworkPrefix netip.Prefix
+			machineNetworkPrefix, err = netip.ParsePrefix(string(machineNetwork.Cidr))
+			if err != nil {
+				return true
+			}
+			return machineNetworkPrefix.Contains(nicIP.Addr())
+		})
+		return found
+	})
```
Comment (eifrach) [inline internal/host/validator.go:1880]: If I understand, you are trying to check if the host IP in the CIDR
let take this part and this check 
https://github.com/openshift/assisted-service/blob/master/internal/network/machine_network_cidr.go#L519
and create a common function to serve them both

make less repeating code

File: internal/host/hostcommands/install_cmd.go [hunk 6]
```diff
@@ -382,20 +382,9 @@ func appendISCSIArgs(installerArgs []string, installationDisk *models.Disk, inve
...
-			return err == nil && iSCSIHostIP.Compare(prefix.Addr()) == 0
-		})
-		return ok
-	})
-
-	if !ok {
-		return nil, fmt.Errorf("Cannot find the interface belonging to iSCSI host IP %s", iSCSIHostIP.String())
+	nic, err := network.FindInterfaceByIP(iSCSIHostIP, inventory.Interfaces)
+	if err != nil {
+		return nil, err
```
Comment (eifrach) [inline internal/host/hostcommands/install_cmd.go:387]: let's keep the error message - helps debuging and better logging 
```suggestion
		return nil, fmt.Errorf("Cannot find the interface belonging to iSCSI host IP %s", iSCSIHostIP.String())
```

File: internal/network/utils_test.go [hunk 7]
```diff
@@ -1012,3 +1013,40 @@ var _ = Describe("GetVips", func() {
...
+			},
+			&searchedInterface,
+		}
+
+	})
+	DescribeTable("FindInterfaceByIP", func(allInterfaces *[]*models.Interface, lookForIp string, expectedInterface *models.Interface, expectedError string) {
+		int, err := FindInterfaceByIPString(lookForIp, *allInterfaces)
+		if expectedError == "" {
+			Expect(err).To(BeNil())
+			Expect(int).To(Equal(expectedInterface))
```
Comment (eifrach) [inline internal/network/utils_test.go:1040]: let's add debug output for the expect command
```suggestion
			Expect(err).To(BeNil(), fmt.Sprintf(err))
			Expect(int).To(Equal(expectedInterface),
			    fmt.Sprintf("expected: %s to be in interface %s", int , expectedInterface ))
```

File: internal/network/utils_test.go [hunk 8]
```diff
@@ -1012,3 +1013,40 @@ var _ = Describe("GetVips", func() {
...
+		} else {
+			Expect(err).ToNot(BeNil())
+			Expect(err.Error()).To(Equal(expectedError))
```
Comment (eifrach) [inline internal/network/utils_test.go:1043]: same

File: internal/host/transition_test.go [hunk 9]
```diff
@@ -6708,6 +6709,19 @@ var _ = Describe("State machine test - refresh transition", func() {
...
+
+			refreshHostArgs.conditions[string(IscsiHostNetworkInterfaceDoesNotBelongToMachineCidr)] = false
+
+			Expect(stateMachine.Run(TransitionTypeRefresh, testState, &refreshHostArgs)).To(Succeed())
+			Expect(string(testState.State())).To(Equal(models.HostStatusInsufficient))
+
+			refreshHostArgs.conditions[string(IscsiHostNetworkInterfaceDoesNotBelongToMachineCidr)] = true
+
+			Expect(stateMachine.Run(TransitionTypeRefresh, testState, &refreshHostArgs)).To(Succeed())
+			Expect(string(testState.State())).To(Equal(models.HostStatusKnown))
```
Comment (eifrach) [inline internal/host/transition_test.go:6723]: same as before let make it a practice to add debug message to Expect .
This will help see the root cause and not some generic failure

**Review by eifrach — CHANGES_REQUESTED** [review]
Great Work, small changes just  to make it easier to read / deubg

I  also want you to add deubg output for the `Expect` I didn't comment on all of them.

File: swagger.yaml (hunk 4 above)
Comment (adriengentil) [reply swagger.yaml:6506]: maybe `iscsi-nic-belong-to-machine-cidr` WDYT?

File: internal/host/validations_test.go (hunk 2 above)
Comment (adriengentil) [reply internal/host/validations_test.go:2581]: MachineNetwork is hard to test because swagger already validates that the field is already a CIDR. So the test don't even comes to this validation.

Comment (adriengentil) [reply internal/host/validations_test.go:2581]: agree that if I may extract a unit test if I split the function

File: internal/host/transition_test.go (hunk 9 above)
Comment (rccrdpccl) [reply internal/host/transition_test.go:6723]: I think it's a great idea to add error expectations, but I think it should be done by type and not by error/debug message. Otherwise tests will be even more fragile (text change will make us fix tests) , and our codebase is already struggling with test fragility.

File: internal/network/utils_test.go (hunk 7 above)
Comment (adriengentil) [reply internal/network/utils_test.go:1040]: the issue is to avoid the truncation ?

File: internal/host/hostcommands/install_cmd.go (hunk 6 above)
Comment (pastequo) [reply internal/host/hostcommands/install_cmd.go:387]: Not sure it's a good idea to lose the original error. Should be wrapped instead

File: internal/network/utils_test.go (hunk 7 above)
Comment (eifrach) [reply internal/network/utils_test.go:1040]: so the message will only be shown if the test will fail -- so it just for debug on failure

File: internal/network/machine_network_cidr.go [hunk 10]
```diff
@@ -516,25 +516,58 @@ func GetDefaultRouteNetworkByFamily(h *models.Host, networks map[AddressFamily][
...
+		}
+		parsedCidr = append(parsedCidr, machineIpnet)
+	}
 
+	return parsedCidr, nil
+}
+
+// Check if a host belongs to all the networks specified as Machine Networks.
+func IsHostInPrimaryMachineNetCidr(log logrus.FieldLogger, cluster *common.Cluster, host *models.Host) bool {
 	// TODO(mko) This rule should be revised as soon as OCP supports multiple machineNetwork
```
Comment (eifrach) [inline internal/network/machine_network_cidr.go:535]: let's check if it's ClusterManagedNetworking retrun true
cause we allow only one network

File: internal/network/utils_test.go (hunk 7 above)
Comment (adriengentil) [reply internal/network/utils_test.go:1040]: I don't understand the point, the test wil already output the diff with the name of the test. Why adding one more output `expected: %s to be in interface %s"` ?

Comment (adriengentil) [reply internal/network/utils_test.go:1040]: just had a look, the report looks like that in case of failure:
```
$ SKIP_UT_DB=1 go test ./internal/network -ginkgo.focus "FindInterfaceByIP"
Running Suite: Machine network cider Suite
==========================================
Random Seed: 1731427695
Will run 5 of 218 specs

SSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSS•
------------------------------
• Failure [0.000 seconds]
FindInterfaceByIP
/home/agentil/Projects/openshift/assisted-service/internal/network/utils_test.go:1016
  FindInterfaceByIP
  /home/agentil/Projects/openshift/assisted-service/internal/network/utils_test.go:1036
    Look for IPv6 [It]
    /home/agentil/Projects/openshift/assisted-service/internal/network/utils_test.go:1047

    Expected
        <*errors.errorString | 0xc000583d10>: 
        Cannot find the network interface on which the IP 1001:db8:0:200::75 is set
        {
            s: "Cannot find the network interface on which the IP 1001:db8:0:200::75 is set",
        }
    to be nil

    /home/agentil/Projects/openshift/assisted-service/internal/network/utils_test.go:1039
------------------------------
•••SSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSS

Summarizing 1 Failure:

[Fail] FindInterfaceByIP FindInterfaceByIP [It] Look for IPv6 
/home/agentil/Projects/openshift/assisted-service/internal/network/utils_test.go:1039

Ran 5 of 218 Specs in 0.001 seconds
FAIL! -- 4 Passed | 1 Failed | 0 Pending | 213 Skipped
--- FAIL: TestMachineNetworkCidr (0.03s)
FAIL
FAIL	github.com/openshift/assisted-service/internal/network	0.105s
FAIL

```

File: internal/network/machine_network_cidr.go (hunk 10 above)
Comment (eifrach) [reply internal/network/machine_network_cidr.go:536]: ignore it - It OK
you can have different host network form ISCSI network

File: internal/network/utils_test.go (hunk 7 above)
Comment (eifrach) [reply internal/network/utils_test.go:1040]: in this case it's not required but, when using `tables` or loops than you don't know what vaules have failed 
but let's take it offline i'll show some examples

File: internal/host/validator.go (hunk 3 above)
Comment (adriengentil) [reply internal/host/validator.go]: it got refactored with functions in network packages

File: internal/host/validator.go (hunk 1 above)
Comment (adriengentil) [reply internal/host/validator.go]: from `netip.ParsePrefix` because `err` is captured in the closure

File: internal/host/transition_test.go (hunk 9 above)
Comment (adriengentil) [reply internal/host/transition_test.go]: let's take it offline

File: internal/network/utils_test.go (hunk 8 above)
Comment (adriengentil) [reply internal/network/utils_test.go:1043]: let's take it offline
